### PR TITLE
Tuto virtual devices

### LINKED ISSUE #448: Tutorial on custom virtual devices
Write a tutorial for the 'Advanced Features' section of the docs explaining:
- The difference between `Device` and `VirtualDevice`
- Why, when and how to create your own custom virtual device 

Should follow #447 .

### PR DESCRIPTION
**Pull request by a-corni**
Creation of tutorial for Devices and Virtual Devices.

It seems to me there is an issue with the method `to_virtual` of `Device`: the virtual device created using this function has the attribute `reusable_channels` set to `False` (as it was set to `False` in `Device`). As it looks like an important property of a virtual device, I have decided to add the possibility to modify its value when converting a device into a virtual device. It is still set to `False` by default yet (this way it passes the former tests). Perhaps should the default value be `True` ? If we agree on the utility of this modification, tests have to be modified and a new test has to be developed.

Further possible developments that emerged by writing the notebook:
- Should not we have a `print_specs` for virtual devices ?
- We still have to define a Rydberg level even in the simplest definition of a `VirtualDevice`. How to know which one to choose ? (I guess it is associated with the wanted value of C6) Perhaps could we set a default value for this as well ?

Closes #448

### CODE REVIEW

**Comment by HGSilveri** [comment]
> It seems to me there is an issue with the method `to_virtual` of `Device`: the virtual device created using this function has the attribute `reusable_channels` set to `False` (as it was set to `False` in `Device`). As it looks like an important property of a virtual device, I have decided to add the possibility to modify its value when converting a device into a virtual device. It is still set to `False` by default yet (this way it passes the former tests). Perhaps should the default value be `True` ? If we agree on the utility of this modification, tests have to be modified and a new test has to be developed.

The idea behind the `to_virtual()` device method was that a `Device` was converted to an identical `VirtualDevice`, so treating the `reusable_channels` parameter specially by allowing it to be modified doesn't make much sense. Once the user has the virtual device, they can create new ones with different parameters with `dataclasses.replace()` (including with `reusable_channels=True`).



> * Should not we have a `print_specs` for virtual devices ?

The main use for `print_specs` is for generating the specs for the physical devices in the docstrings. It is cumbersome to maintain and adds little value outside of that (especially when you consider that dataclasses already come with decent `__repr__()` methods built-in).

> * We still have to define a Rydberg level even in the simplest definition of a `VirtualDevice`. How to know which one to choose ? (I guess it is associated with the wanted value of C6) Perhaps could we set a default value for this as well ?

This is a good point, I guess the user will be left a bit clueless on this in some cases. I'm not against setting a default value, I'm just not sure there is a single good candidate

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 1]
```diff
@@ -0,0 +1,321 @@
+{
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:1]: <p>We usually put <code>()</code> when referring to a method (and often include the class to be unequivocal), so I would replace <code>to_virtual</code> with <code>Device.to_virtual()</code>    </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/451/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:1]: <p>Maybe add a subtitle so that the <code>replace()</code> feature is highlighted</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/451/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:1]: <ul><li>The way you put it, it seems like the only things we can change is <code>rydberg_level</code> and <code>reusable_channels</code>, which is not the case. </li><li>You should hide the link in an hyperlink   </li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/451/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:1]: <p>I would maybe not use DQC, just so it's not confused with the DQC protocol.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/451/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:1]: <p>Remember that devices are generally immutable, so directly linking your variable to a built-in device is not a good idea. I would suggest you use <code>replace</code> again to show you can start from <code>MockDevice</code> and make a few tweaks to have your own mock device.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/451/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

**Review by HGSilveri — COMMENTED** [review]
It's looking good, but I have an overall remark: in some sections, the text is written as if it is just a series of bullet points, it doesn't flow so well and makes it sound slightly robotic. 
I would suggest you take a look to the existing tutorials to get a sense of the tone we are going for with these.

File: pulser-core/pulser/devices/_device_datacls.py [hunk 2]
```diff
@@ -621,6 +621,10 @@ def _optional_parameters(self) -> tuple[str, ...]:
     def change_rydberg_level(self, ryd_lvl: int) -> None:
         """Changes the Rydberg level used in the Device.
 
+        Find the $C_6$ coefficient matching the Rydberg level on
+        `this page <https://github.com/pasqal-io/Pulser/blob/develop/
+        pulser-core/pulser/devices/interaction_coefficients/C6_coeffs.json>`
+
```
Comment (HGSilveri) [inline pulser-core/pulser/devices/_device_datacls.py:627]: Hum, this didn't work out unfortunately...
image

**Review by HGSilveri — COMMENTED** [review]
Can you please add the tutorial to the docs so that we can check if it renders correctly?

**Comment by HGSilveri** [comment]
Hey @a-corni , is this ready for another review?

**Comment by HGSilveri** [comment]
I noticed that they are on the docs now, but you should clear the outputs so that they are executed when the docs are built

**Comment by a-corni** [comment]
It was ready for review, but let me clear the outputs first.

De : Henrique Silvério ***@***.***>
Envoyé : vendredi 13 janvier 2023 14:26
À : pasqal-io/Pulser ***@***.***>
Cc : Antoine CORNILLOT ***@***.***>; Mention ***@***.***>
Objet : Re: [pasqal-io/Pulser] Tuto virtual devices (PR #451)


I noticed that they are on the docs now, but you should clear the outputs so that they are executed when the docs are built

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+   "metadata": {},
+   "source": [
+    "## Physical devices"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To perform any computation using Pulser, it is necessary to define a device. Some physical devices are available, they can be accessed via `pulser.devices`. These devices are instances of the `Device` class. They are constrained by physical considerations and all their parameters are defined. Below is presented the specifications of the physical device `Chadoq2`, that are accessible via the `Device.print_specs()` method. "
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:34]: Here are a few suggestions, feel free to adopt the ones you agree with:
```suggestion
    "To perform any computation using Pulser, it is necessary to choose a device. For convenience, some examples of typical physical devices are included and can be accessed via `pulser.devices`.\n
     These devices are instances of the `Device` class. By definition, a `Device` represents a physical device and, as such, requires that all its parameters are defined. \n
     As an example, we present below the specifications of the physical device `Chadoq2`, which can be accessed via the `Device.print_specs()` method. "
```

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+   "metadata": {},
+   "source": [
+    "### Converting a Device into a Virtual Device"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "However we sometimes want to perform the computations on a more permissive device, a device that would have more dimensions or more atoms for instance, or more types of channels. This can be done on an emulator behaving like a device. The `VirtualDevice` class is useful to define such an emulator, a virtual device. Let's start by configuring a virtual device having the same parameters as `Chadoq2`. To do this, we use the `Device.to_virtual()` method that creates a virtual device from a physical one. "
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:65]: For readability, it usually helps not to put everything on the same paragraph
```suggestion
    "However we sometimes want to perform the computations on a more permissive device, a device that would have more dimensions or more atoms for instance, or more types of channels. This can be done on an emulator behaving like a device. The `VirtualDevice` class is useful to define such an emulator, a virtual device.\n 
    Let's start by configuring a virtual device having the same parameters as `Chadoq2`. To do this, we use the `Device.to_virtual()` method that creates a virtual device from a physical one. "
```

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+   "metadata": {},
+   "source": [
+    "### Building your own virtual device"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "More generally, the `VirtualDevice` class is more permissive than the `Device` class. Actually, some parameters can be left undefined and simpler devices can be built with a `VirtualDevice` instance. A virtual device only needs a `name`, a `dimension` and a `rydberg_level` to be initialized."
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:159]: nit:
```suggestion
    "More generally, the `VirtualDevice` class is more permissive than the `Device` class. In fact, some parameters can be left undefined and simpler devices can be built with a `VirtualDevice` instance. A virtual device only needs a `name`, a `dimension` and a `rydberg_level` to be initialized."
```

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+   "metadata": {},
+   "source": [
+    "### Defining the channels of your device"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Nevertheless, to perform computation, channels have to be defined. The channels enabled on the device are defined in `channel_objects`. Their name can be defined in `channel_id`. If no name is provided, the name of the channels are automatically generated as `channeltype_adressing`. For example, the `Rydberg.Global` channel is automatically named `rydberg_global` in the specifications of `Chadoq2`."
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:187]:
```suggestion
    "Nevertheless, to perform computations, channels have to be defined. The channels enabled on the device are defined in `channel_objects`. Their IDs can be defined in `channel_ids`, but if no IDs are provided, they will be automatically generated as `{channeltype}_addressing`. For example, the `Rydberg.Global` channel is automatically named `rydberg_global` in the specifications of `Chadoq2`."
```

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# This basic device can be used for digital quantum computing\n",
+    "DiQCVirtualDevice = replace(\n",
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:197]: How about `DigitalQCVirtualDevice`? It's okay to have a long name here imo

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+    "    ),\n",
+    ")\n",
+    "print(DiQCVirtualDevice)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### An example of virtual device: the `MockDevice`"
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:215]:
```suggestion
    "### An built-in example of a virtual device: the `MockDevice`"
```

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 9]
```diff
@@ -79,40 +55,32 @@
...
-     ]
-    }
-   ],
+   "outputs": [],
    "source": [
     "# Converting the Device object in a VirtualDevice object\n",
     "VirtualChadoq = Chadoq2.to_virtual()\n",
-    "print(VirtualChadoq)"
+    "VirtualChadoq.__repr__()"
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:76]: Why this change? I'm okay with using the `repr`, but in that case I would prefer `repr(VirtualChadoq)` instead

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 10]
```diff
@@ -79,40 +55,32 @@
...
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "### Changing parameters of a virtual device with `replace()`"
+    "### Changing parameters of a virtual device with `dataclass.replace()`"
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:83]:
```suggestion
    "### Changing parameters of a virtual device with `dataclasses.replace()`"
```

**Review by HGSilveri — COMMENTED** [review]
Looking good! I only have a handful of suggestion at this point

**Comment by a-corni** [comment]
Are we happy with this output of VirtualDevices ? Should I change __repr__ ?
I could insert `\n` after each argument or copy the `Device.to_repr()` method ?
image

File: tutorials/advanced_features/Virtual Devices.ipynb (hunk 3 above)
Comment (HGSilveri) [reply tutorials/advanced_features/Virtual Devices.ipynb:34]: By the way, the formatting on these suggestions is probably screwed up, so don't just commit them even if you agree with everything

**Comment by HGSilveri** [comment]
> Are we happy with this output of VirtualDevices ? Should I change `__repr__` ? I could insert `\n` after each argument or copy the `Device.to_repr()` method ? image

Have you tried doing just `VirtualDevice`, without the `print` or the `repr`? From experience, that usually formats better on notebooks

**Comment by a-corni** [comment]
I don't see any improvement...

image

**Comment by HGSilveri** [comment]
> I don't see any improvement...
> 
> image

Yeah, it might be an issue with how notebooks are rendered on sphinx. I was expecting something like this: 
image

Considering it works properly on normal notebooks, I would say it's not worth it to go and change the `__repr__()`

**Comment by HGSilveri** [comment]
But, if you're up for it, you can try playing with [this parameter](https://nbsphinx.readthedocs.io/en/latest/usage.html#nbsphinx_responsive_width), it seems like it might do what you want

**Comment by a-corni** [comment]
I have tried various things last friday without managing to solve the issue. I will try to come back to it later, one possible solution I have read could be to disable the "scrolling" option. Another thing to do is to compare the configuration of our notebooks with the configuration of nbsphynx. My guess is that a default value is different between the two of them that explains the difference in display.

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+   "metadata": {},
+   "source": [
+    "### Converting a Device into a Virtual Device"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "However we sometimes want to perform the computations on a more permissive device, a device that would have more dimensions or more atoms for instance, or more types of channels. This can be done on an emulator behaving like a device. The `VirtualDevice` class is useful to define such an emulator, a virtual device.\n",
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:69]:
```suggestion
    "However, we sometimes want to perform the computations on a more permissive device, a device that would have more dimensions or more atoms for instance, or more types of channels. This can be done on an emulator behaving like a device. The `VirtualDevice` class is useful to define such an emulator, a virtual device.\n",
```

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To perform any computation using Pulser, it is necessary to choose a device. For convenience, some examples of typical physical devices are included and can be accessed via `pulser.devices`.\n",
+    "\n",
+    "These devices are instances of the `Device` class. They are constrained by physical considerations and all their parameters are defined.\n",
+    "\n",
+    "As an exemple, we present below the specifications of the physical device `Chadoq2`, which can be accessed via the `Device.print_specs()` method."
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:38]:
```suggestion
    "As an example, we present below the specifications of the physical device `Chadoq2`, which can be accessed via the `Device.print_specs()` method."
```

**Review by HGSilveri — APPROVED** [review]
Just some final remarks, but overall LGTM!

File: docs/source/_static/css/no_autoscroll.css [hunk 13]
```diff
@@ -0,0 +1,3 @@
+pre {
```
Comment (HGSilveri) [inline docs/source/_static/css/no_autoscroll.css:1]: This might not work simply because you didn't add the path to this file to `html_css_files` in `conf.py`

File: docs/source/conf.py [hunk 14]
```diff
@@ -74,3 +74,13 @@
...
+
+nbsphinx_prolog = """
+.. raw:: html
+
+    <style>
+        pre {
+            white-space: pre-wrap;
+        }
+    </style>
+"""
```
Comment (HGSilveri) [inline docs/source/conf.py:86]: If this is not working, maybe remove it?

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 15]
```diff
@@ -9,7 +9,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 1,
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:12]: This is preventing the tutorial from being executed

File: docs/source/_static/css/no_autoscroll.css (hunk 13 above)
Comment (HGSilveri) [reply docs/source/_static/css/no_autoscroll.css:1]: You might want to remove this file too

File: tutorials/advanced_features/Virtual Devices.ipynb [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+    "As an example, we present below the specifications of the physical device `Chadoq2`, which can be accessed via the `Device.print_specs()` method."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Dipslay Chadoq2's specifications\n",
```
Comment (HGSilveri) [inline tutorials/advanced_features/Virtual Devices.ipynb:46]:
```suggestion
    "# Display Chadoq2's specifications\n",
```

**Comment by HGSilveri** [comment]
The tutorial is still not building...
